### ISSUE TITLE
Panic "decoding unset EncDatum" on SELECT * FROM JOIN

### ISSUE THREAD

**Issue by fastest963** [cockroachdb/cockroach#22647]
Here are the first few lines of the logs, I don't see anything with `[config]`.
```
I180213 17:12:22.962846 1 cli/start.go:890  CockroachDB CCL v2.0-alpha.20180212 (linux amd64, built 2018/02/12 20:00:23, go1.9.3)
I180213 17:12:23.064417 1 server/config.go:313  available memory from cgroups (8.0 EiB) exceeds system memory 7.3 GiB, using system memory
I180213 17:12:23.064488 1 server/config.go:411  system total memory: 7.3 GiB
I180213 17:12:23.064562 1 server/config.go:413  server configuration:
max offset             500000000
cache size             1.8 GiB
SQL memory pool size   1.8 GiB
scan interval          10m0s
scan max idle time     200ms
event log enabled      true
I180213 17:12:23.064614 1 cli/start.go:772  process identity: uid 0 euid 0 gid 0 egid 0
I180213 17:12:23.064638 1 cli/start.go:459  starting cockroach node
I180213 17:12:23.070982 12 storage/engine/rocksdb.go:544  opening rocksdb instance at "/cockroach/cockroach-data/cockroach-temp129578226"
I180213 17:12:23.100374 12 storage/engine/rocksdb.go:544  opening rocksdb instance at "/cockroach/cockroach-data"
I180213 17:12:23.116274 12 server/config.go:518  [n?] 1 storage engine initialized
I180213 17:12:23.116332 12 server/config.go:521  [n?] RocksDB cache size: 1.8 GiB
I180213 17:12:23.116484 12 server/config.go:521  [n?] store 0: RocksDB, max size 0 B, max open file limit 1043576
W180213 17:12:23.120411 12 gossip/gossip.go:1292  [n?] no incoming or outgoing connections
I180213 17:12:23.125931 12 server/server.go:963  [n?] sleeping for 478.28414ms to guarantee HLC monotonicity
I180213 17:12:23.195009 14 gossip/client.go:129  [n?] started gossip client to 10.24.6.8:26257
```

When running the following query:
```
SELECT * FROM sessions JOIN users USING (region, uuid) WHERE (sessions.region = 'gce-us-central1') AND (sessions.id = '0u00013eea105611e884b1cb8831b4b057') AND (sessions.uuid = '0u00013eea105611e8869bd975a668d5cf');
```
on the tables:
```
CREATE TABLE IF NOT EXISTS users (
	region STRING,
	uuid   STRING,
	lp     STRING,
	lfv    INT,
	PRIMARY KEY (region, uuid)
)
```
```
CREATE TABLE IF NOT EXISTS sessions (
	region STRING,
	uuid   STRING,
	id     STRING,
	fv     INT,
	lv 	   INT,
	nv 	   INT,
	lda    INT,
	f      INT,
	sf     INT,
	b      BOOL,
	hab    BOOL,
	u      STRING,
	ip     STRING,
	ua     STRING,
	PRIMARY KEY (region, uuid, id),
	CONSTRAINT fk_user FOREIGN KEY (region, uuid) REFERENCES users
) INTERLEAVE IN PARENT users (region, uuid)
```

Causes the node to panic and crash:
```
panic: decoding unset EncDatum

goroutine 4613 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*EncDatum).EnsureDecoded(0xc426dea120, 0xc437e06540, 0xc437e02758, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:198 +0x502
github.com/cockroachdb/cockroach/pkg/sql.(*distSQLReceiver).Push(0xc437e026c0, 0xc426dea030, 0x10, 0x10, 0x0, 0x12)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:423 +0x695
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).EmitRow(0xc426dea000, 0x350e0c0, 0xc424f11080, 0xc426de8700, 0x12, 0x12, 0x4, 0xc426de8700, 0x12)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:275 +0x145
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*interleavedReaderJoiner).Run(0xc426dea000, 0xc420639870)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/interleaved_reader_joiner.go:359 +0xafd
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:466 +0x3c5
```

Doing `SELECT users.*` works fine though, but as soon as you try and do `*` with `sessions` it panics.

Also if I switch the query around to being:
```
SELECT * FROM users JOIN sessions
```
then it doesn't panic.

**Comment by fastest963** [comment]
I'm not sure if this is related but when I run:
```
SELECT s.id AS id, u.* FROM sessions AS s JOIN users AS u ON ((s.region=u.region) AND (s.uuid=u.uuid)) WHERE (s.region='gce-us-central1');
```
I get:
```
pq: did not find terminator 0x0 in buffer 0x126763652d75732d63656e7472616c310002
```

So maybe my dataset is in some weird state?

**Comment by arjunravinarayan** [comment]
@fastest963 thank you for the bug report! I'm not sure its related, do you might filing a second bug for that as well? Just to make sure we address all the issue you might be facing.

cc @solongordon - does this look related at all to the other panic you were looking at?

**Comment by fastest963** [comment]
No problem! Made a second bug for the terminator error.

**Comment by solongordon** [comment]
@arjunravinarayan Just from looking at the stacktrace it doesn't look related to the other panic.

**Comment by BramGruneir** [comment]
@pbardea's steps to repro the panic:
```sql
INSERT INTO users VALUES ('gce-us-central1', '0u00013eea105611e8869bd975a668d5cf', 'qwer', 123);

INSERT INTO sessions VALUES ('gce-us-central1', '0u00013eea105611e8869bd975a668d5cf', '0u00013eea105611e884b1cb8831b4b057', 123, 123, 123, 123, 123, 123, true, true, 'abc', 'abc', 'abc');

SELECT * FROM sessions JOIN users USING (region, uuid) WHERE (sessions.region = 'gce-us-central1') AND (sessions.id = '0u00013eea105611e884b1cb8831b4b057') AND (sessions.uuid = '0u00013eea105611e8869bd975a668d5cf');
```

Just confirmed it locally.

**Comment by BramGruneir** [comment]
Fixed in #23229.
